tiledb/sm/buffer/buffer.h

```diff
@@ -276,6 +303,148 @@ class Buffer {
   Status ensure_alloced_size(uint64_t nbytes);
 };
 
+// =======================================================
+//      ConstBuffer
+// =======================================================
+/**
+ * A read-only buffer fully-initialized at construction. It does not manage
+ * memory; its storage is subordinate to some other object.
+ */
+class ConstBuffer : public BufferBase {
+ public:
+  /** Default constructor deleted. */
+  ConstBuffer() = delete;
+
+  /**
+   * Ordinary constructor.
+   *
+   * @param data The data of the buffer.
+   * @param size The size of the buffer.
+   */
+  ConstBuffer(const void* data, uint64_t size);
+
+  /**
+   * Constructor from a general-purpose buffer.
+   *
+   * @param buff The buffer the object will encapsulate, working on its
+   *     data and size, but using a separate local offset, without affecting
+   *     the input buffer.
+   */
+  explicit ConstBuffer(Buffer* buff);
+
+  /** Destructor. */
+  ~ConstBuffer() = default;
+
+  /** Returns the buffer data. */
+  const void* data() const;
+
+  /** Returns pointer to data at the current offset. */
+  const void* cur_data() const;
+
+  /** Returns the number of bytes left for reading. */
+  uint64_t nbytes_left_to_read() const;
+
+  /**
+   * This is a special function used for reading from a buffer that stores
+   * uint64_t values. It reads a number of bytes from the local buffer and
+   * writes them to the input buffer, after adding *offset* to each read
+   * uint64_t value.
+   *
+   * @param buf The buffer to write to when reading from the local buffer.
+   * @param nbytes The number of bytes to read.
+   * @param offset The offset to add to each uint64_t value.
+   * @return void.
+   */
+  void read_with_shift(uint64_t* buf, uint64_t nbytes, uint64_t offset);
+
+  /**
+   * Returns a value from the buffer of type T.
+   *
+   * @tparam T The type of the value to be read.
+   * @param offset The offset in the local buffer to read from.
+   * @return The desired value of type T.
+   */
+  template <class T>
+  inline T value(uint64_t offset) const {
+    return ((const T*)(((const char*)data_) + offset))[0];
+  }
```

Comment:
A)Should this be in BufferBase?
B) Would an assert() checking the value of 'offset' parameter be a good idea?

Reply:
It could be put there, but it's not being used there, so I didn't move it. The scope of this PR isn't to do all possible refactoring on these classes, but enough to fix the defects.

And congratulation, you found I defect I missed. 'offset' could indeed be > size here.

Reply:
Added assertions.

Reply:
Added assertion.